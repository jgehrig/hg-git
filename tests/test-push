#!/bin/sh

# Fails for some reason, need to investigate
# "$TESTDIR/hghave" git || exit 80

# bail if the user does not have dulwich
python -c 'import dulwich, dulwich.repo' || exit 80

# bail early if the user is already running git-daemon
echo hi | nc localhost 9418 2>/dev/null && exit 80

echo "[extensions]" >> $HGRCPATH
echo "hggit=$(echo $(dirname $(dirname $0)))/hggit" >> $HGRCPATH
echo 'hgext.graphlog =' >> $HGRCPATH
echo 'hgext.bookmarks =' >> $HGRCPATH


GIT_AUTHOR_NAME='test'; export GIT_AUTHOR_NAME
GIT_AUTHOR_EMAIL='test@example.org'; export GIT_AUTHOR_EMAIL
GIT_AUTHOR_DATE="2007-01-01 00:00:00 +0000"; export GIT_AUTHOR_DATE
GIT_COMMITTER_NAME="$GIT_AUTHOR_NAME"; export GIT_COMMITTER_NAME
GIT_COMMITTER_EMAIL="$GIT_AUTHOR_EMAIL"; export GIT_COMMITTER_EMAIL
GIT_COMMITTER_DATE="$GIT_AUTHOR_DATE"; export GIT_COMMITTER_DATE

count=10
commit()
{
    GIT_AUTHOR_DATE="2007-01-01 00:00:$count +0000"
    GIT_COMMITTER_DATE="$GIT_AUTHOR_DATE"
    git commit "$@" >/dev/null 2>/dev/null || echo "git commit error"
    count=`expr $count + 1`
}
hgcommit()
{
    HGDATE="2007-01-01 00:00:$count +0000"
    hg commit -d "$HGDATE" "$@" >/dev/null 2>/dev/null || echo "hg commit error"
    count=`expr $count + 1`
}

mkdir gitrepo
cd gitrepo
git init | python -c "import sys; print sys.stdin.read().replace('$(dirname $(pwd))/', '')"

echo alpha > alpha
git add alpha
commit -m "add alpha"
git checkout -b not-master 2>&1 | sed s/\'/\"/g

# dulwich does not presently support local git repos, workaround
cd ..
git daemon --base-path="$(pwd)"\
 --listen=localhost\
 --export-all\
 --pid-file="$DAEMON_PIDS" \
 --detach --reuseaddr \
 --enable=receive-pack

hg clone git://localhost/gitrepo hgrepo | grep -v '^updating'

cd hgrepo
echo beta > beta
hg add beta
hgcommit -m 'add beta'


echo gamma > gamma
hg add gamma
hgcommit -m 'add gamma'

hg book -r 1 beta
hg push -r beta

cd ..

echo % should have two different branches
cd gitrepo
git branch -v

echo % some more work on master from git
git checkout master 2>&1 | sed s/\'/\"/g
echo delta > delta
git add delta
commit -m "add delta"
git checkout not-master 2>&1 | sed s/\'/\"/g

cd ..

cd hgrepo
echo % this should fail
hg push -r master

echo % ... even with -f
hg push -fr master

hg pull
# TODO shouldn't need to do this since we're (in theory) pushing master explicitly,
# which should not implicitly also push the not-master ref.
hg book not-master -r default/not-master --force
echo % master and default/master should be diferent
hg log -r master | grep -v ': *master'
hg log -r default/master

echo % this should also fail
hg push -r master

echo % ... but succeed with -f
hg push -fr master

cd ..
