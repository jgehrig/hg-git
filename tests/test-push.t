Load commonly used test logic
  $ . "$TESTDIR/testutil"

  $ git init gitrepo
  Initialized empty Git repository in $TESTTMP/gitrepo/.git/
  $ cd gitrepo
  $ echo alpha > alpha
  $ git add alpha
  $ fn_git_commit -m "add alpha"
  $ git checkout -b not-master 2>&1 | sed s/\'/\"/g
  Switched to a new branch "not-master"

  $ cd ..
  $ hg clone gitrepo hgrepo | grep -v '^updating'
  importing git objects into hg
  1 files updated, 0 files merged, 0 files removed, 0 files unresolved

  $ cd hgrepo
  $ echo beta > beta
  $ hg add beta
  $ fn_hg_commit -m 'add beta'


  $ echo gamma > gamma
  $ hg add gamma
  $ fn_hg_commit -m 'add gamma'

  $ hg book -r 1 beta
  $ hg push -r beta
  pushing to $TESTTMP/gitrepo
  searching for changes
  adding objects
  added 1 commits with 1 trees and 1 blobs
  adding reference refs/heads/beta

  $ cd ..

should have two different branches
  $ cd gitrepo
  $ git branch -v
    beta       cffa0e8 add beta
    master     7eeab2e add alpha
  * not-master 7eeab2e add alpha

some more work on master from git
  $ git checkout master 2>&1 | sed s/\'/\"/g
  Switched to branch "master"
  $ echo delta > delta
  $ git add delta
  $ fn_git_commit -m "add delta"
  $ git checkout not-master 2>&1 | sed s/\'/\"/g
  Switched to branch "not-master"

  $ cd ..

  $ cd hgrepo
this should fail
  $ hg push -r master
  pushing to $TESTTMP/gitrepo
  searching for changes
  abort: branch 'refs/heads/master' changed on the server, please pull and merge before pushing
  [255]

... even with -f
  $ hg push -fr master
  pushing to $TESTTMP/gitrepo
  searching for changes
  abort: branch 'refs/heads/master' changed on the server, please pull and merge before pushing
  [255]

  $ hg pull 2>&1 | grep -v 'divergent bookmark'
  pulling from $TESTTMP/gitrepo
  importing git objects into hg
  (run 'hg update' to get a working copy)
TODO shouldn't need to do this since we're (in theory) pushing master explicitly,
which should not implicitly also push the not-master ref.
  $ hg book not-master -r default/not-master --force
master and default/master should be diferent
  $ hg log -r master | grep -v ': *master'
  changeset:   2:72f56395749d
  user:        test
  date:        Mon Jan 01 00:00:12 2007 +0000
  summary:     add gamma
  
  $ hg log -r default/master | grep -v 'master@default'
  changeset:   3:1436150b86c2
  tag:         default/master
  tag:         tip
  parent:      0:3442585be8a6
  user:        test <test@example.org>
  date:        Mon Jan 01 00:00:13 2007 +0000
  summary:     add delta
  

this should also fail
  $ hg push -r master
  pushing to $TESTTMP/gitrepo
  searching for changes
  abort: pushing refs/heads/master overwrites 72f56395749d
  [255]

... but succeed with -f
  $ hg push -fr master
  pushing to $TESTTMP/gitrepo
  searching for changes
  adding objects
  added 1 commits with 1 trees and 1 blobs
  updating reference refs/heads/master

this should fail, no changes to push
The exit code for this was broken in Mercurial (incorrectly returning 0) until
issue3228 was fixed in 2.1
  $ hg push -r master && false
  pushing to $TESTTMP/gitrepo
  searching for changes
  no changes found
  [1]

  $ cd ..

Push empty Hg repo to empty Git repo (issue #58)
Since there aren't any changes, exit code 1 is expected in modern Mercurial.
However, since it varies between supported Mercurial versions, we need to
force it to consistency for now. (see issue3228, fixed in Mercurial 2.1)
  $ hg init hgrepo2
  $ git init -q --bare gitrepo2
  $ hg -R hgrepo2 push gitrepo2 && false
  pushing to gitrepo2
  searching for changes
  no changes found
  [1]
