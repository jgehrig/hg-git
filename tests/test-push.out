Initialized empty Git repository in gitrepo/.git/

Switched to a new branch "not-master"
importing git objects into hg
1 files updated, 0 files merged, 0 files removed, 0 files unresolved
pushing to git://localhost/gitrepo
exporting hg objects to git
creating and sending data
    default::refs/heads/beta => GIT:cffa0e8d
    default::refs/heads/not-master => GIT:7eeab2ea
    default::refs/heads/master => GIT:7eeab2ea
% should have two different branches
  beta       cffa0e8 add beta
  master     7eeab2e add alpha
* not-master 7eeab2e add alpha
% some more work on master from git
Switched to branch "master"
Switched to branch "not-master"
% this should fail
pushing to git://localhost/gitrepo
creating and sending data
abort: refs/heads/master changed on the server, please pull and merge before pushing
% ... even with -f
pushing to git://localhost/gitrepo
creating and sending data
abort: refs/heads/master changed on the server, please pull and merge before pushing
pulling from git://localhost/gitrepo
importing git objects into hg
(run 'hg update' to get a working copy)
% master and default/master should be diferent
changeset:   2:72f56395749d
user:        test
date:        Mon Jan 01 00:00:12 2007 +0000
summary:     add gamma

changeset:   3:1436150b86c2
tag:         default/master
tag:         tip
parent:      0:3442585be8a6
user:        test <test@example.org>
date:        Mon Jan 01 00:00:13 2007 +0000
summary:     add delta

% this should also fail
pushing to git://localhost/gitrepo
creating and sending data
abort: pushing refs/heads/master overwrites 72f56395749d
% ... but succeed with -f
pushing to git://localhost/gitrepo
creating and sending data
    default::refs/heads/beta => GIT:cffa0e8d
    default::refs/heads/not-master => GIT:7eeab2ea
    default::refs/heads/master => GIT:cc119202
